Enable Low Power State/System Off State on nrf52 platform

youvedeep opened the pull request:
This patch series implements Power Management for nrf52 series SOC. 
This composed of multiple patches :- 
1. **gpio: Add support for SENSE configuration for nrf SOC GPIO.** ==> This is to configure the GPIO in sense mode. Currently the GPIO can not be programmed into Sense mode on Zephyr. This patch address this issue.
2. **gpio: Add APIs for enabling/Disabling GPIOTE interrupts for nrf52 SOC.**  ==> For waking up the device from Low Power mode via GPIO, we need to enable the interrupt detect at PORT. (INTENSET.PORT). So this patch adds APIs to set GPIOTE event.
3. **ARM: nrf52: Power Management for nrf52 series SOC** ==> This patch adds SOC specif functions for invoking Power Management (Low Power Mode (CONSTLAT and LOWPWR tasks) and System OFF modes).
4. **samples: boards: nrf52: test demo for nrf52 power management** ==> This patch adds sample application for testing power management for nrf52 series SOCs

youvedeep-singh committed:
gpio: Add support for SENSE configuration for nrf SOC GPIO.

nrf SOCs are capable of waking from Low power state or
Deep Sleep state using sense configuration.
So adding support for this in nrf GPIO driver.

Jira: ZEP-2623

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

tbursztyka commented on include/gpio.h:
```diff
@@ -189,8 +189,29 @@ extern "C" {
 
+
+
+#if defined(CONFIG_SOC_FAMILY_NRF5)
```
gpio.h is meant to be generic among all SoC/Boards, so you cannot add SoC specific features there.

nashif replied:
thoe values need to come either from the SDK or should be declared in the SOC related files. See ext/hal/nordic/mdk, for example grep for GPIO_PIN_CNF_SENSE_Low

carlescufi commented:
@anangl can you take a look at this?

youvedeep-singh committed:
gpio: Add APIs for enabling/Disabling GPIOTE interrupts for nrf52 SOC.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
ARM: nrf52: Power Management for nrf52 series SOC

Add support for nrf52 series SOC. This patch Adds :-
1. Architecture specific Power Management APIs.
2. APIs for invoking various Power Management tasks into nrf52.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
samples: boards: nrf52: test demo for nrf52 power management

This sample tesecase provides sample code :-
1. Triggers Low Power tasks into nrf52 SOC (CONST LAT and LOW PWR)
2. Triggers System Off state.
3. Wake up device through port Event by GPIO (Button 1).

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+.. _power-mgr-sample:
```
/samples/boards/quark_se_c1000/power_mgr/README.rst has this same label (labels must be unique across all documents).  Maybe call this "nrf52-power-mgr-sample"?

dbkinder replied:
Because this sample is down one more directory level than the others in samples/boards, please also make a change to get rid of the message saying this document isn't included in any toctree.  In samples/boards/boards.rst, change the line that says:

     */*

to

     **/*

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+Overview
+********
+
+A sample implementation of a power manager app that uses the Zephyr
```
Make this a sentence by changing the start:

     This sample demonstrates a power manager app ...

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+A sample implementation of a power manager app that uses the Zephyr
+power management infrastructure.
+
+This app will cycle through the various power schemes at every call
```
How about:

     This app cycles through the following power state schemes each time the
     _sys_soc_suspend() hook function is called:

dbkinder replied:
Is it really a "cycle", or does it stop after three calls?

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+
+1. CPU Low Power State
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
+   task on nrf SOC. In this mode Low Exit Latency. In this Mode 16MHz Clock is
```
These lines should be indented to preserve the second-level list you've started.  You'll also need a blank line before the start of the indented list:

     1. CPU Low Power State

        A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTANT
           task on nrf SOC. In this mode Low Exit Latency. ...

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+It will cycle through the following states:
+
+1. CPU Low Power State
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
```
Is the word CONSTLAT correct? (or CONSTANT misspelled :)

youvedeep replied:
as per nrf52 reference manual (Page 80, 18.3.1 Sub power modes) , they refer it as CONSTLAT task only. So CONSTLAT is correct.

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
+   task on nrf SOC. In this mode Low Exit Latency. In this Mode 16MHz Clock is
+   kept on.
+   B. LOW POWER STATE : In this Power State CONSTLAT task is triggered on nrf
```
CONSTLAT again

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+   task on nrf SOC. In this mode Low Exit Latency. In this Mode 16MHz Clock is
+   kept on.
+   B. LOW POWER STATE : In this Power State CONSTLAT task is triggered on nrf
+   SOC. This is deeper power state than CONSTANT LATENCY Mode, In this mode
```
is a deeper

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+   SOC. This is deeper power state than CONSTANT LATENCY Mode, In this mode
+   16MHz Clock is turned off and only 32KHz clock is retained.
+
+2. Deep Sleep - This Power State is mapped to System OFF state. In this Mode
```
In this mode all devices on board are suspended and the SOC
     and all devices lose context on wake up.

Capital letters should be lowercase.

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+   All devices on board gets suspended and SOC also looses context. So All
+   devices and SOC loses context on wake up.
+
+3. No-op - no operation and letting kernel do its idle
```
no operation, letting the kernel idle

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,58 @@
+.. code-block:: console
+
+    make BOARD=nrf52_pca10040
+
```
you showed how to build, but not how to run

dbkinder reviewed (changes requested):
suggested edits for clarity, and getting rid of doc gen errors

youvedeep commented:
Thanks for review @dbkinder , I will incorporate these changes in next version of patch.

youvedeep commented:
Addressed comments from @dbkinder .

carlescufi commented:
@nvlsianpu @anangl could you take a look at this PR to enable System OFF?

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+
+void configure_ram_retention(void)
+{
+	/* Place Holder for RAM retainition */
```
retention

carlescufi replied:
Does the kernel currently have support for retaining state over a full power cycle like the one we see with System OFF? If so it would be great if we could use it as an example

youvedeep replied:
No It does not support retaining the state on full cycle off. It loses context and it is starting from fresh boot. 
My intention was to start a patch, Later on we can enhance the functionality to save context in some memory location and restore the context on fresh boot.

youvedeep replied:
Currently on issuing System_Off state it just turns off. And on pressing GPIO it starts from fresh boot.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/soc_power.h:
```diff
@@ -0,0 +1,47 @@
+
+/*
+ * Bit 0 from GP0 register is used internally by the kernel
+ * to handle PM multicore support. Any change on QMSI and/or
```
I don't think we should mention QMSI in a Nordic file?

youvedeep replied:
Yes i missing removing this, I copied this portion from Curie, I will remove this,

carlescufi commented on ext/hal/nordic/hal/nrf_power.h:
```diff
@@ -40,7 +40,7 @@
  */
 #include "nrf.h"
-#include "nordic_common.h"
+//#include "nordic_common.h"
```
why? also, commented-out code should not be part of the PR

youvedeep replied:
Thanks for catching this, I missed this, I will remove this.

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+   Low Power State is SOC specific state, In state is transparent to devices.
+   SOC and devices do not lose context in this Mode. SOC supports following
+   two Low Power states :-
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
```
needs a blank line before this line that's starting a list, also, remove the ``-`` at the end of the previous line

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+=============
+
+nrf52 core output
+---------------
```
underline for title needs to be at least as long as the title (it can be longer but not shorter)

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+power management infrastructure to enter into Low Power state.
+
+This app will cycle through following power schemes each time idle thread calls
+_sys_soc_suspend() hook function :-
```
remove the ``-`` at the end of this line

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+This app will cycle through following power schemes each time idle thread calls
+_sys_soc_suspend() hook function :-
+
+1. Low Power State:-
```
remove ``-`` at the end of the line

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+      SOC. This is a deeper power state than CONSTANT LATENCY Mode. In this mode
+      16MHz Clock is turned off and only 32KHz clock is retained.
+
+2. Deep Sleep:- This Power State is mapped to System OFF state. In this mode
```
just use a ``:`` and not ``:-``

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+This sample demonstrares a power manager app that uses the Zephyr
+power management infrastructure to enter into Low Power state.
+
+This app will cycle through following power schemes each time idle thread calls
```
through the following

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+
+1. Low Power State:-
+   Low Power State is SOC specific state, In state is transparent to devices.
+   SOC and devices do not lose context in this Mode. SOC supports following
```
supports the following

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+   SOC and devices do not lose context in this Mode. SOC supports following
+   two Low Power states :-
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
+      task on nrf52 SOC. In this mode Low Exit Latency. In this Mode 16MHz Clock
```
In this mode there is Low Exit latency and the 16MHz clock is kept on.

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+      16MHz Clock is turned off and only 32KHz clock is retained.
+
+2. Deep Sleep:- This Power State is mapped to System OFF state. In this mode
+   all devices on board gets suspended. All devices and SOC loses context on
```
get suspended

     lose context

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,73 @@
+
+.. code-block:: console
+
+    Building : make BOARD=nrf52_pca10040
```
This would read better as three separate code blocks and putting the words that aren't command (Building, Flashing, and Running) outside the code block.  For example:

     Flashing (Use nrfjprog for flashing):

     .. code-block:: console

        nrfjprog --eraseall -f nrf52
        nrfjprog --program <zephyr.hex> -f nrf52
        nrfjprog --reset -f nrf52
 
     Running:

     .. code-block:: console

dbkinder reviewed (changes requested):
Getting closer. Thanks!

nvlsianpu commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	/* Wait for __WFE*/
+	_issue_low_power_command();
```
I just read in OPS 20.2 Port event:
> Putting the system into System ON IDLE while DETECT is high will not cause DETECT to wake the system
up again.

@anang I;m not sure whether we need a check/prevent this just  before lpw cmd is issued?

youvedeep replied:
Yes I got your point, But I am not getting any reference, how we can see if DETECT is high or low. 
My understanding is - 
Detect will remain High for the duration till GPIO is Low (incase configured Low as in our Case). https://devzone.nordicsemi.com/question/171620/clear-gpio-sense-signal-on-hitolo/

So i think we need to check our application if GPIO is high and allow attempting Low Power state if GPIO is high. What is your suggestion on this ?

youvedeep replied:
Guys looks like our merge window is approaching, Can you please provide comments on this. 

Basic purpose of this patch is to have basic skeleton for nrf Low power state and Deep sleep state. If any enhancement (like restore context on wake up) is to be done we can take that up later ?

nvlsianpu replied:
@youvedeep 
> So i think we need to check our application if GPIO is high and allow attempting Low Power state if GPIO is high.

OK - we can improve this in future PR if we will want.

nvlsianpu replied:
But it will be great if add a remark(comment) near the sleep call execution.
When you will add assertions i requested this patch will be ready for me.

youvedeep replied:
Thanks @nvlsianpu  I will add assertion :), 

Also good idea, 
We will check GPIO status before attempting next sleep entry so that will make sure DETECT is Low.

youvedeep replied:
Also I was thinking of one more change. 

Currently I have am calling "configure_ram_retention(void)" inside SOC specific file. But i feel it should be left to application developer which RAM blocks she want to configure based on his application. So I am thinking of moving that to Application, just before entertaining into Low Power states.

nvlsianpu replied:
Agree on that - the setup of Backups-regs, Ram retention etc. should be a separate operation.

youvedeep commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+		break;
+#endif
+	default:
+		/* Unsupported State */
```
Sure I will put one print message indicating this as "Not a valid state".

youvedeep commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+		break;
+
+	default:
+		/* Unsupported State */
```
Sure I will put one print message indicating this as "Not a valid state".

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+/* System_OFF is deepest Power state available, On exiting from this
+ * state CPU including all peripherals reset
+ */
+static void _system_off(void)
```
This function becomes unused when `CONFIG_SYS_POWER_DEEP_SLEEP` is not defined.
Thus, it should be surrounded with `#if (defined(CONFIG_SYS_POWER_DEEP_SLEEP))`.

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+		break;
+	case SYS_POWER_STATE_MAX:
+		break;
+	case SYS_POWER_STATE_DEEP_SLEEP_1:
```
_DEEP_SLEEP_1, _DEEP_SLEEP_2, and _MAX shouldn't be listed here, since by definition they won't need any post operations (first two are unsupported, the last one is a "do nothing").
But is there a point in providing all these cases when each one is empty? Wouldn't a comment like "Currently, there is no need to perform any specific operations after exiting low power states." be enough instead?
Currently, there iis no need to do any post ops

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+	nrf_power_system_off();
+}
+
+void configure_ram_retention(void)
```
`static` missing

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,94 @@
+	/* Place Holder for RAM retention */
+}
+
+void _issue_low_power_command(void)
```
`static` missing

anangl commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -175,6 +182,10 @@ static int gpio_nrf5_config(struct device *dev,
 
+		if (sense == GPIO_SENSE_INVALID) {
+			sense = GPIO_SENSE_DISABLE;
+		}
```
An assertion would be better than such silent correction. User should be notified that the passed `flags` value is not valid.

anangl commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+   SOC supports the following two Low Power states :
+
+   A. CONSTANT LATENCY LOW POWER MODE : This Low Power State triggers CONSTLAT
+      task on nrf52 SOC. In this mode Low Exit Latency. In this Mode there is Low
```
"In this mode Low Exit Latency." seems to be remains of a previous version.

youvedeep replied:
Thanks for pointing out, When i modified it while addressing the comments, I made this typo. I will change it,

anangl commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -0,0 +1,315 @@
+#define DEVICE_POLICY_MAX 15
+static char device_ordered_list[DEVICE_POLICY_MAX];
+static char device_retval[DEVICE_POLICY_MAX];
+int pm_state;
```
`static` missing

youvedeep replied:
I will change it to static.

anangl commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -0,0 +1,315 @@
+	printk("====Deep sleep exit!\n");
+}
+
+static int low_power_state_entry(int pm_state)
```
This parameters overlays the same named global variable. And it seems to be unneeded.

youvedeep replied:
I kept it same while keeping uniformity with other power management sample application .(samples/boards/nrf52/power_mgr/src/main.c)

I got your point, Sure I will modify it.

anangl commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -0,0 +1,315 @@
+
+	if ((ticks != K_FOREVER) && (ticks < MIN_TIME_TO_SUSPEND)) {
+		printk("Not enough time for PM operations "
+				"TIME_UNIT_STRING : %d).\n", ticks);
```
`TIME_UNIT_STRING` was probably intended to be outside of the string literal.

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/soc_power.h:
```diff
@@ -0,0 +1,39 @@
+	SYS_POWER_STATE_CPU_LPS_2,     /* Not supported*/
+	SYS_POWER_STATE_DEEP_SLEEP,    /* System Off */
+	SYS_POWER_STATE_DEEP_SLEEP_1,  /* Not Supported */
+	SYS_POWER_STATE_DEEP_SLEEP_2,  /* Not Supported */
```
Perhaps I'm missing some context here. What is the purpose of defining states that are not (and will not be) supported? Is the intention to make this enumerated type common for all the chips? But why is it then defined separately in the same form but in several places?

youvedeep replied:
I have keep consistency across the Chips.

youvedeep replied:
We are planning to have a "power.h" file which is same across architectures/Chips. So this duplication  of these states for different SOCs will be gone. This is for stop gap arrangement.

youvedeep commented:
Addressed comments from @anangl .

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+Overview
+********
+
+This sample demonstrares a power manager app that uses the Zephyr
```
demonstrates

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+      is kept ON.
+   B. LOW POWER STATE : In this Power State CONSTLAT task is triggered on nrf52
+      SOC. This is a deeper power state than CONSTANT LATENCY Mode. In this mode
+      16MHz Clock is turned off and only 32KHz clock is retained.
```
the 16MHz clock is turned off and only the 32KHz clock is used.

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+      16MHz Clock is turned off and only 32KHz clock is retained.
+
+2. Deep Sleep: This Power State is mapped to SYSTEM OFF state. In this mode
+   all devices on board gets suspended. All devices and SOC lose context on
```
get

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+Building, Flashing and Running
+******************************
+Building
+    .. code-block:: console
```
delete the leading spaces before ``.. code-block``

dbkinder replied:
delete leading spaces before ``.. code-block``

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+
+Running:
+
+   .. code-block:: console
```
delete leading spaces before ``.. code-block``

dbkinder replied:
delete this line (what's block is not a code block)

dbkinder commented on samples/boards/nrf52/power_mgr/README.rst:
```diff
@@ -0,0 +1,78 @@
+
+   .. code-block:: console
+
+    1. Open UART terminal.
```
delete the leading spaces before each of these numbered list items

dbkinder reviewed (changes requested):
suggested edits

andrewboie commented:
@dbkinder can you please use the "start a review" feature instead of leaving many one-shot comments, it will greatly reduce the amount of github email spam

dbkinder commented:
Of course! Sorry, I didn't know "start a review" would prevent that.

-- david kinder

On Oct 19, 2017, at 3:53 PM, Andrew Boie <notifications@github.com<mailto:notifications@github.com>> wrote:


@dbkinder<https://github.com/dbkinder> can you please use the "start a review" feature instead of leaving many one-shot comments, it will greatly reduce the amount of github email spam

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/1546#issuecomment-338060030>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKjRQtKpaxR0i_y5V_XKMq8-21iu1-lTks5st9LwgaJpZM4PcDj5>.

youvedeep commented:
Addressed comments from @dbkinder.

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,106 @@
+	__SEV();
+	__WFE();
+	__WFE();
+	if (NRF_GPIOTE->EVENTS_PORT) {
```
Move this to line 61 in the caller of this function, and use nRFx interface if available.

youvedeep replied:
Hi @cvinayak , 


I did not find any interface available for clearing NRF_GPIOTE->EVENTS_PORT. 
Clearing the port Event is needed when sense is enabled and port event is generated. In other cases like RTC wake. reset button we will not need this, But it should not create harm in other cases.


My understanding (through reading some nrf52 spec and same applications @ https://github.com/NordicSemiconductor/nrf51-powerdown-examples) is when we are using port Event for waking up the device then we will need to clear the port event after waking. So it will be needed for IDLE ON States too. In-fact in case of System Off state device is going to reset, so we may not need this.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,106 @@
+void _issue_low_power_command(void)
+{
+	/* Issue _WFE Enent*/
+	__SEV();
```
Do you want to sleep unconditionally? Because this will do this, regardless of whether an interrupt was already pending. If you want to sleep only if an interrupt is *not* pending then you should do:

__WFE();
__SEV();
__WFE();

carlescufi replied:
Just to further clarify this. If you  use the SEV/WFE/WFE sequence then  you will go to sleep even if there was an interrupt executed between the entry to this function and the second WFE. Not sure if that is what we want, since in that case that interrupt may have done something that the kernel needs to react to. The other option I suggested, WFE/SEV/WFE, will only go to sleep if there is no interrupts pending (first WFE), and then the pend bit will be cleared unconditionally (SEV/WFE)

youvedeep replied:
good Idea @carlescufi, we will change to WFE/SEV/WFE.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,106 @@
+
+void _issue_low_power_command(void)
+{
+	/* Issue _WFE Enent*/
```
Event

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -98,8 +100,13 @@ struct gpio_nrf5_data {
 
-#define GPIO_SENSE_DISABLE    (GPIO_PIN_CNF_SENSE_Disabled << GPIO_PIN_CNF_SENSE_Pos)
-#define GPIO_SENSE_ENABLE     (GPIO_PIN_CNF_SENSE_Enabled << GPIO_PIN_CNF_SENSE_Pos)
+#define GPIO_SENSE_DISABLE	(GPIO_PIN_CNF_SENSE_Disabled	\
```
Indentation and line split, do it after an operator:

#define GPIO_SENSE_DISABLE	(GPIO_PIN_CNF_SENSE_Disabled << \
                                                         GPIO_PIN_CNF_SENSE_Pos)

anangl commented on arch/arm/soc/nordic_nrf5/nrf52/power.c:
```diff
@@ -0,0 +1,106 @@
+}
+#endif
+
+void _issue_low_power_command(void)
```
This function should be `static`.

anangl commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -175,6 +183,12 @@ static int gpio_nrf5_config(struct device *dev,
+		if (sense == GPIO_SENSE_INVALID) {
+			SYS_LOG_ERR("Invalid parameter for sense,"
+						"setting SENSE disable\n");
+			sense = GPIO_SENSE_DISABLE;
```
I thought that you'd rather use an __ASSERT, similarly to what is done a couple of lines above with `ds_low` and 'ds_high`. Such setting of SENSE is simply invalid, and if it is used in an application, it means that there is obviously an error in it. I think the driver should not try to correct it.

anangl commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -0,0 +1,330 @@
+
+	if ((ticks != K_FOREVER) && (ticks < MIN_TIME_TO_SUSPEND)) {
+		printk("Not enough time for PM operations "
+				"%s : %d).\n", TIME_UNIT_STRING, ticks);
```
TIME_UNIT_STRING is constant, so the same effect you'll achieve with:
`TIME_UNIT_STRING": %d).\n", ticks);`
and the generated code will be simpler.

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -375,6 +383,28 @@ static const struct gpio_driver_api gpio_nrf5_drv_api_funcs = {
+}
+
+/* Disable GPIOTE Interrupt */
+void  nrf_gpiote_interrupt_disable(uint32_t mask)
```
There is no reference to this function, delete it.

youvedeep replied:
as we have option to Enable GPIOTE interrupt, I want to put for Disable too for completeness incase someone want to use it. 
If we feel it is not needed, i can put another patch to delete it.

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -98,8 +97,11 @@ struct gpio_nrf5_data {
 
-#define GPIO_SENSE_DISABLE    (GPIO_PIN_CNF_SENSE_Disabled << GPIO_PIN_CNF_SENSE_Pos)
-#define GPIO_SENSE_ENABLE     (GPIO_PIN_CNF_SENSE_Enabled << GPIO_PIN_CNF_SENSE_Pos)
+#define GPIO_SENSE_DISABLE	(GPIO_PIN_CNF_SENSE_Disabled <<	\
```
Tabs used between definition, and next define is not aligned with other defines before and after.

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -375,6 +383,28 @@ static const struct gpio_driver_api gpio_nrf5_drv_api_funcs = {
 };
 
+/* Enable GPIOTE Interrupt */
+void nrf_gpiote_interrupt_enable(uint32_t mask)
```
This function is not needed (and use of extern in caller c files). Use nrf_gpiote_int_enable instead and just a function (if not already exists) to clear EVENT, called before nrf_gpiote_int_enable is called by caller code.

youvedeep replied:
nrf_gpiote_interrupt_enable() is called by application code whereas nrf_gpiote_int_enable is present as a HAL function (static inline inside ext/hal/nordic/hal/nrf_gpiote.h).  It is not recommended to call directly as HAL function from sample application. 
Also this puts sort of abstraction layer for applications developer so that she need not worry in case HAL is changing later point.

cvinayak replied:
ok, agree. No change required here.

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -375,6 +383,28 @@ static const struct gpio_driver_api gpio_nrf5_drv_api_funcs = {
+/* Clear GPIOTE Port Event */
+void nrf_gpiote_clear_port_event(void)
+{
+	if (NRF_GPIOTE->EVENTS_PORT) {
```
its optimal to just set to 0 (always), than have a branch operations (if-clause). Check if you can reuse code to clear EVENT, i.e. any nrf_gpiote_* function that clears EVENT.

youvedeep replied:
Good point, we can remove the condition. I had put it as it was put in nrf sample application. But i agree we can remove it.

Also i do not see any reference to function in HAL to clear GPIOTE Event.

cvinayak replied:
The sample is managing LED. But in this PR, the if-clause is not needed, please remove.

cvinayak commented on samples/boards/nrf52/power_mgr/src/main.c:
```diff
@@ -0,0 +1,330 @@
+				| GPIO_PUD_PULL_UP
+				| GPIO_CFG_SENSE_LOW);
+	/* Enable GPIOTE Port Event */
+	nrf_gpiote_interrupt_enable(GPIOTE_INTENSET_PORT_Msk);
```
I prefer not to hide the EVENT clear inside this function, refer to my other comments related to use if this function.

youvedeep commented:
Addressed comments from @cvinayak

carlescufi commented:
@nashif @tbursztyka @dbkinder Looks like everything is fine on the Nordic side now if you want to revisit this
